### PR TITLE
Add test to consensus builder

### PR DESCRIPTION
**Pull request by arrivets**
Test that builder does not include receipts whose parent result is
not incorporated in the chain or in the mempool.

### CODE REVIEW

File: module/builder/consensus/builder_test.go
Comment (zhangchiqing) [inline module/builder/consensus/builder_test.go:815]: I could think of a few more cases to cover, but I'm not more if they have been covered

denote `B(r_A)` as block B contains a receipt for block A; X is the new block to build

## 1. receipts for blocks on other forks should not be included
```
A <- B(r_A) <- C <- X
  ^- D(r_A) <- E
```
receipts (r_D) should NOT be included in `X`, because it's for block on different fork

## 2. receipts for the same block but different execution paths should be included
denote `B'` is a receipt with different execution result from `B`
```
A <- B(r_A) <- C(r_B, r_B') <- X
```
receipts `r_C, r_C'` should be included in `X`, because they are valid extension of their own execution path: `r_B <- r_C`, `r_B' <- r_C'`.

## 3. receipts extends a result on a different fork should be excluded
```
A <- B(r_A) <- C(r_B) <- X
            ^- D(r_B')
```
Among receipts `r_C, r_C'`, only `r_C` should be included. `r_C'` should not be included, because it extends from `r_B'`, which only exists on  a different fork.

## 4. receipts exists on the same fork should be excluded
```
A <- B(r_A) <- C(r_B) <- X
```

receipt `r_B` should be excluded, because it's already exist in `C`


## 5. receipts exists on a different fork should be included
```
A <- B(r_A) <- C(r_B) <- X
            ^- D(r_B')
```

receipt `r_B'` should be included, because the fork `A <- B <- C` doesn't have this receipt yet.

File: module/builder/consensus/builder_test.go
Comment (zhangchiqing) [inline module/builder/consensus/builder_test.go:766]: maybe we could extract a function so that this can be simplified into `bs.storageHasBlock(a)`

File: module/builder/consensus/builder_test.go
Comment (zhangchiqing) [inline module/builder/consensus/builder_test.go:747]: FYI. I made a function that can [prepare such chain much easier](https://github.com/onflow/flow-go/pull/320/files#diff-77f09607e896b942749ffed547ac64d32d9007b8b564b86bd0c942dd7d063b0bR262-R272)

Comment (arrivets) [reply module/builder/consensus/builder_test.go:751]: Nice. Perhaps it makes sense that I target this PR at your branch: leo/validate-multi-receipts
Or I can wait until yours is merged

Comment (zhangchiqing) [reply module/builder/consensus/builder_test.go:751]: yup, my PR has been merged to master

**Comment by AlexHentschel** [comment]
I have added the remaining unit tests for the Execution Tree. By now, the tests for the consensus builder are broken up over two components:
* The [`execution_tree_test.go`](https://github.com/onflow/flow-go/blob/aca053f1905bd77656597827fac643604b97f9d6/module/mempool/consensus/execution_tree_test.go) holds the unit tests for the `ExecutionTree`
* The builder tests [`builder_test.go`](https://github.com/onflow/flow-go/blob/aca053f1905bd77656597827fac643604b97f9d6/module/builder/consensus/builder_test.go) tests whether the `Builder` is supplying the correct control parameters (`BlockFilter` and `ReceiptFilter`) to the `ExecutionTree` (mocked out in the testes)

By induction, it _should_ follow that `Builder` with a full `ExecutionTree` functions correctly. @zhangchiqing and I synced last week and we concluded that this is a pretty large assumption, given the complexity of both components. We therefore thought, it would be great to implement the [tests Leo outlined](https://github.com/onflow/flow-go/pull/322#discussion_r564128362) as "integration tests". 

I have updated the implementation of the first test according to [Leo's suggestion](https://github.com/onflow/flow-go/pull/322#discussion_r564131513). I also updated the naming slightly: https://github.com/onflow/flow-go/blob/aca053f1905bd77656597827fac643604b97f9d6/module/builder/consensus/builder_test.go#L814 to reflect that it is an integration test. 

@arrivets 
* it would be great if you could implement the remaining integration tests 2, ..., 4 according to [Leo's comment](https://github.com/onflow/flow-go/pull/322#discussion_r564128362). 
* Regarding test 3, the expected behaviour would be to include  `r_C` _as well as_ `r_B'` and `r_C`. In other words:
   * all known results for the fork `A <- B <- C` are `r_A, r_B, r_B', r_C, r_C'` 
   * but since we have already included `r_A, r_B,` in ancestor blocks, only `r_B', r_C, r_C'` remain to be included
* I believe test 5 is a simplified version of the scenario that we are checking in test 3 (@zhangchiqing does that sound convincing?).

File: module/builder/consensus/builder_test.go
Comment (zhangchiqing) [inline module/builder/consensus/builder_test.go:923]: yes, this is the right behavior. Thanks for the fix

**Review by zhangchiqing — APPROVED** [review]
Great tests! 💯

File: module/mempool/consensus/execution_tree_test.go
Comment (durkmurder) [inline module/mempool/consensus/execution_tree_test.go:181]: Suggestion: first check that there is no path and after adding bridge verify that there is a path

**Review by durkmurder — APPROVED** [review]
LGTM

File: module/mempool/consensus/execution_tree_test.go
Comment (AlexHentschel) [reply module/mempool/consensus/execution_tree_test.go:181]: 👍 added in commit https://github.com/onflow/flow-go/pull/322/commits/2ed4e73f56bea54f3339053be8660d7123b82039

**Review by AlexHentschel — APPROVED** [review]
thanks Martin. Great work. 
🚢

**Pull request merged**
